### PR TITLE
Support for source host in mysql static configuration

### PR DESCRIPTION
**Pull request by rafael**
### Description 
* MySQL supports accounts per host name: `CREATE USER 'david'@'localhost'`
  This will enforce that this account only works when connecting from
  localhost.

* The following PR adds a starting point for this feature in Vitess. The main
  changes are:
    - Added new config format to support multiple user credentials
      depending on the `SourceHost` when using mysql static auth.
    - Backwards compatibility for config.
    - Updated `auth_server_static` to match both user and source host when
      performing authentication. At the moment, only `""` or `localhost`
      are supported. Empty string means it could connect from everywhere.
    - Added support to start a listener in a unix socket. When a config entry has
      `SourceHost` equal to localhost, it will only be allowed to connect via this 
      socket.
* This initial approach is not a complete port of MySQL feature, as it doesn't 
   add support for full wildcards, host match and network masks as the source 
   hosts.  If we see need, in the future we could match the MySQL 
   implementation.

### Usage

* When starting a vtgate you can pass this new flag: `-mysql_server_socket_path`. When present, it will listen for connections in that unix socket path. 
* In addition, when specifying an static config entry, you can provide `SourceHost` with `localhost`. This entry will only work when connecting via the unix socket (same behavior as you would have with MySQL).
* From a mysql client this will work: 
  ```
  $ mysql -h localhost -uuname -p  # Yei! 
  ```

### Tests

* So far I have tested these changes in a local cluster and our development environment.

### CODE REVIEW

**Comment by sougou** [comment]
@alainjobart this looks like a good addition.

File: go/vt/vtgate/plugin_mysql_server.go [hunk 1]
```diff
@@ -158,11 +160,21 @@ func initMySQLProtocol() {
...
 		mysqlListener.SlowConnectWarnThreshold = *mysqlSlowConnectWarnThreshold
 	}
-
-	// And starts listening.
+	// Start listening for tcp
 	go func() {
 		mysqlListener.Accept()
 	}()
+
+	if mysqlServerSocketPath != nil {
```
Comment (alainjobart) [inline go/vt/vtgate/plugin_mysql_server.go]: This should be:
if *mysqlServerSocketPath != "" {
...

File: go/vt/vtgate/plugin_mysql_server.go [hunk 2]
```diff
@@ -117,6 +118,7 @@ func (vh *vtgateHandler) ComQuery(c *mysql.Conn, query []byte, callback func(*sq
 }
 
 var mysqlListener *mysql.Listener
+var mysqlUnixListener *mysql.Listener
 
 // initiMySQLProtocol starts the mysql protocol.
 // It should be called only once in a process.
```
Comment (alainjobart) [inline go/vt/vtgate/plugin_mysql_server.go]: Under there, the next if should probably be:
if *mysqlServerPort < 0  && *mysqlServerSocketPath != "" {
return
}

That way it's possible to have only the socket file, and no port.

Then, line 142 when it says:
// Create a listener
that entire block should be with an:
if *mysqlServerPort >= 0 {
up to the end of the TCP socket stuff.

**Comment by alainjobart** [comment]
LGTM I'll let @sougou merge when convenient.

**Pull request merged**